File: 20221024_Monday/20221024.rb
# Leetcode: 1239. Maximum Length of a Concatenated String with Unique Characters.
# https://leetcode.com/problems/maximum-length-of-a-concatenated-string-with-unique-characters/
# @param {String[]} arr
# @return {Integer}
def max_length(arr)
  @d = true
  @max = 0
  # Filter string with duplicates.
  @a = arr.select{|x| x.chars.size == x.chars.uniq.size}
  @masks = @a.map{|x| str2mask(x)}
  @lens = @a.map(&:size)
  (0...@a.size).each do |i|
    dfs(i,0,0,"")
  end
  return @max
end

def dfs(i,total_mask, total_size, str)
  puts "#{i} #{@a[i]}"
  total_mask |= @masks[i]
  total_size += @lens[i]
  str = str + @a[i]
  if @max < total_size
    @max = total_size
    p [str,total_size] if @d
  end
  (i+1...@a.size).each do |j|
    check = 0 == (total_mask & @masks[j])
    puts "i=#{i} j=#{j} str=#{str} check=#{check.inspect}" if @d
    puts "#{str_mask(total_mask)} : #{str}" if @d
    puts "#{str_mask(@masks[j])} : #{@a[j]}" if @d
    puts "#{str_mask(total_mask & @masks[j])} : #{str + "&" + @a[j]}" if @d
    # Fix! Need to have j instead of i.
    dfs(j,total_mask, total_size, str) if 0 == (total_mask & @masks[j]) 
  end
end

def str_mask(mask)
  mask.to_s(2).rjust(26,?0)
end

def str2mask(str)
  mask = 0
  str.chars.each do |x|
    # Fix! Forget to move one.
    mask = mask | (1 << (x.ord - ?a.ord)) # Fix!
  end
  return mask
end

File: 20221024_Monday/20221024.go
// Leetcode: 1239. Maximum Length of a Concatenated String with Unique Characters.
// https://leetcode.com/problems/maximum-length-of-a-concatenated-string-with-unique-characters/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 0 ms, faster than 100.00% of Go online submissions for Maximum Length of a Concatenated String with Unique Characters.
// Memory Usage: 2.3 MB, less than 68.42% of Go online submissions for Maximum Length of a Concatenated String with Unique Characters.
var max int
var a *[]string // I suppose it could be solved without a pointer. Just by a slice.
var masks *[]int
var lens *[]int
func maxLength(arr []string) int {
  max = 0
  // Filter uniq.
  tmp := make([]string,0)
  for _,str := range arr{
    if uniq(str){
      tmp = append(tmp,str)
    }
  }
  a = &tmp
  tmp_masks := make([]int,len(*a))
  tmp_lens := make([]int,len(*a))
  for i,str := range *a {
    tmp_masks[i] = get_mask(str) 
    tmp_lens[i] = len(str)
  }
  masks = &tmp_masks
  lens = &tmp_lens
  for i:=0; i < len(*a); i+=1{
    dfs(i,0,0)
  }

  return max
}
func dfs(i int, total_mask int, total_len int){
  m,l := *masks, *lens
  total_mask |= m[i]
  total_len += l[i]
  if max < total_len {
    max = total_len
  }
  for j:= i+1; j < len(*a); j += 1{
    if 0 == (total_mask & m[j]){
      dfs(j, total_mask, total_len)
    }
  }
}

func uniq(str string) bool{
  counter := make(map[rune]int)
  for _,ch := range str {
    counter[ch] += 1
    if 2 == counter[ch] {
      return false
    }
  }
  return true
}
func get_mask(str string) int{
  mask := 0
  for _,ch := range str{
    mask = mask | (1 << (ch - 97))
  }
  return mask
}
